### PR TITLE
Fill logics into CustomEquipmentCraft action

### PR DESCRIPTION
**Pull request by U-lis**
Add logic for CustomEquipmentCraft action.

### Review Point
- Do we have plenty of test cases? (missing edge case?)
- Is it good to generate and give substat/skill adding logic?

### CODE REVIEW

**Comment by U-lis** [comment]
idun setting 을 위해 사후 리뷰로 넘깁니다 cc. @ipdae

File: Lib9c/Action/CustomEquipmentCraft/CustomEquipmentCraft.cs [hunk 1]
```diff
@@ -6,20 +6,26 @@
...
 {
     [Serializable]
-    [ActionType(ActionTypeIdentifier)]
+    [ActionType(TypeIdentifier)]
     public class CustomEquipmentCraft : ActionBase
     {
-        public const string ActionTypeIdentifier = "custom_equipment_craft";
+        public const string TypeIdentifier = "custom_equipment_craft";
+        public const int DrawingItemId = 600401;
+        public const int DrawingToolItemId = 600402;
```
Comment (ipdae) [inline Lib9c/Action/CustomEquipmentCraft/CustomEquipmentCraft.cs:27]: 이건 별도 MaterialItemSheet를 사용하지 않는걸까요? 추가로 해머에선 놓쳤는데 이 아이템들은 별도의 ItemSubType을 지정해두는게 좋을것 같습니다.

File: Lib9c/Action/CustomEquipmentCraft/CustomEquipmentCraft.cs [hunk 2]
```diff
@@ -126,22 +139,141 @@ public override IWorld Execute(IActionContext context)
             }
             // ~Validate Recipe ResultEquipmentId
 
-            // Validate Materials
-            // Should finalize cost using sheet, additional cost and proficiency
+            // Modify cost to get real cost
+            var requiredFungibleItems = new Dictionary<int, int>();
```
Comment (ipdae) [inline Lib9c/Action/CustomEquipmentCraft/CustomEquipmentCraft.cs:143]: 아래까지의 코드를 봐선 우선 필요 재료들을 모아둔 다음 removeMaterial은 loop안에서 처리하는게 코드를 파악하기 용이할것같습니다.

File: Lib9c/TableData/CustomEquipmentCraft/CustomEquipmentCraftRelationshipSheet.cs [hunk 3]
```diff
@@ -8,18 +8,19 @@ namespace Nekoyume.TableData.CustomEquipmentCraft
...
         [Serializable]
         public class Row : SheetRow<int>
         {
-            public override int Key => Proficiency;
+            public override int Key => Relationship;
 
-            public int Proficiency { get; private set; }
+            public int Relationship { get; private set; }
             public decimal CostMultiplier { get; private set; }
+            public decimal RequiredBlockMultiplier { get; private set; }
```
Comment (ipdae) [inline Lib9c/TableData/CustomEquipmentCraft/CustomEquipmentCraftRelationshipSheet.cs:21]: 이건 타입이 decimal이어야할 이유가 있는걸까요? 코드를 봐선 long타입이 더 맞는것같습니다.

Comment (ipdae) [reply Lib9c/TableData/CustomEquipmentCraft/CustomEquipmentCraftRelationshipSheet.cs:21]: 다른 multiplier들도 가능하면 decimal대신 만분율을 적용한다던지 하는식으로 단순화시키는게 좋을것같습니다.

File: Lib9c/Model/Item/ItemFactory.cs [hunk 4]
```diff
@@ -126,6 +133,146 @@ public static TradableMaterial CreateTradableMaterial(MaterialItemSheet.Row row)
...
+            foreach (var opt in optionSheet.Values
+                         .Where(row => row.ItemSubType == itemSubType))
+            {
+                optionSelector.Add(opt, opt.Ratio);
+            }
+
+            return optionSelector.Select(1).First();
+        }
+
+        public static Skill.Skill SelectSkill(
```
Comment (ipdae) [inline Lib9c/Model/Item/ItemFactory.cs:200]: 위치가 SkillFactory에 있는게 더 알맞을것 같습니다.

File: Lib9c/Action/CustomEquipmentCraft/CustomEquipmentCraft.cs [hunk 5]
```diff
@@ -126,22 +139,141 @@ public override IWorld Execute(IActionContext context)
...
+            var skill = ItemFactory.SelectSkill(
+                recipeRow.ItemSubType,
+                sheets.GetSheet<CustomEquipmentCraftRecipeSkillSheet>(),
+                sheets.GetSheet<EquipmentItemOptionSheet>(),
+                sheets.GetSheet<SkillSheet>(),
+                random
+            );
+
+            // Create equipment with ItemFactory
+            var equipment = ItemFactory.CreateCustomEquipment(
```
Comment (ipdae) [inline Lib9c/Action/CustomEquipmentCraft/CustomEquipmentCraft.cs:229]: CreateCustomEquipment의 경우를 제외하고 위에서 장비 옵션관련으로 불리는 메서드들이 사용되는곳이 있을까요? 만약 없다면 관련 처리들은 이 메서드 안에서 해도 될것같습니다.

File: Lib9c/Battle/CPHelper.cs [hunk 6]
```diff
@@ -226,6 +226,40 @@ public static decimal GetStatCP(StatType statType, decimal statValue, int charac
...
+                    return cp / characterLevel / 20m;
+                case StatType.CDMG:
+                    return cp / characterLevel / 3m;
+                case StatType.ArmorPenetration:
+                    return cp / 5m;
+                case StatType.Thorn:
+                    return cp / 1m;
+                case StatType.NONE:
+                default:
+                    // throw new ArgumentOutOfRangeException(nameof(statType), statType, null);
```
Comment (ipdae) [inline Lib9c/Battle/CPHelper.cs:258]: 이 주석은 지워도 될듯합니다.

File: Lib9c/Action/CustomEquipmentCraft/CustomEquipmentCraft.cs (hunk 2 above)
Comment (U-lis) [reply Lib9c/Action/CustomEquipmentCraft/CustomEquipmentCraft.cs:143]: 네 안그래도 비용 계산을 helper 로 빼기로 해서 같이 정리될 것으로 보입니다.

File: Lib9c/Action/CustomEquipmentCraft/CustomEquipmentCraft.cs (hunk 1 above)
Comment (U-lis) [reply Lib9c/Action/CustomEquipmentCraft/CustomEquipmentCraft.cs:27]: MaterialItemSheet 에 있기는 한데 ItemId 가 고정이어서 넣어놨습니다. 어차피 sheet 에서 불러도 저 ID 로 부를거라 차이 없을거라 생각했습니다.
+ 도면이랑 제도도구의 subtype 이 두개 생겨야 할까요? 사실 이 subtype 의 용처가 확 와닿지 않아서..

File: Lib9c/TableData/CustomEquipmentCraft/CustomEquipmentCraftRelationshipSheet.cs (hunk 3 above)
Comment (U-lis) [reply Lib9c/TableData/CustomEquipmentCraft/CustomEquipmentCraftRelationshipSheet.cs:21]: 1.2배, 1.5배 같은걸 생각했는데, 만분율은 논의해보겠습니다.

**Comment by U-lis** [comment]
여기 있는 review 내역에 대한 반영은 별도 PR 로 진행합니다.
